File: vignettes/19_declarative/python/19_declarative.py
"""
Declarative Agents (Python)

This sample demonstrates how to define agents and workflows declaratively
using YAML configuration files, then load and run them with AgentFactory
and WorkflowFactory. It mirrors the Julia vignette 19_declarative.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-declarative agent-framework-ollama
"""

import asyncio
from pathlib import Path

from agent_framework import tool
from agent_framework.declarative import AgentFactory, WorkflowFactory
from agent_framework.ollama import OllamaChatClient
from pydantic import Field
from typing import Annotated


# ── A. Agent from YAML string ──────────────────────────────────────────── #

WEATHER_AGENT_YAML = """\
kind: Prompt
name: WeatherHelper
description: A weather assistant that answers concisely
instructions: You answer weather questions concisely.
"""


async def agent_from_yaml_string() -> None:
    """Create an agent from an inline YAML string."""
    print("=== A. Agent from YAML String ===\n")

    factory = AgentFactory(
        client=OllamaChatClient(
            host="http://localhost:11434",
            model="qwen3:8b",
        ),
        safe_mode=False,
    )

    agent = factory.create_agent_from_yaml(WEATHER_AGENT_YAML)
    response = await agent.run("Is it sunny in Paris today?")
    print(f"Agent: {response.text}\n")


# ── B. Agent from file ─────────────────────────────────────────────────── #


async def agent_from_file() -> None:
    """Load an agent definition from a YAML file on disk."""
    print("=== B. Agent from File ===\n")

    yaml_path = Path(__file__).parent / "agents" / "weather.yaml"
    if not yaml_path.exists():
        print(f"  Skipped — {yaml_path} not found.\n")
        return

    factory = AgentFactory(
        client=OllamaChatClient(
            host="http://localhost:11434",
            model="qwen3:8b",
        ),
        safe_mode=False,
    )
    agent = factory.create_agent_from_yaml_path(yaml_path)
    response = await agent.run("Will it rain in Tokyo tomorrow?")
    print(f"Agent: {response.text}\n")


# ── C. Registering tools for declarative use ────────────────────────────── #


@tool(approval_mode="never_require")
def get_weather(
    location: Annotated[str, Field(description="The location to get the weather for.")],
) -> str:
    """Get the current weather for a location."""
    from random import choice, randint
    conditions = ["sunny", "cloudy", "rainy", "stormy"]
    return f"The weather in {location} is {choice(conditions)} at {randint(10, 30)}°C."


TOOL_AGENT_YAML = """\
kind: Prompt
name: WeatherAgent
description: Uses tools to answer weather questions
instructions: Use the available tools to answer weather questions accurately.
tools:
  - kind: function
    name: get_weather
    description: Get the current weather for a location.
    parameters:
      properties:
        location:
          kind: string
          description: The location to get the weather for.
          required: true
"""


async def agent_with_tools() -> None:
    """Create a declarative agent with function tools via bindings."""
    print("=== C. Agent with Registered Tools ===\n")

    factory = AgentFactory(
        client=OllamaChatClient(
            host="http://localhost:11434",
            model="qwen3:8b",
        ),
        bindings={"get_weather": get_weather},
        safe_mode=False,
    )

    agent = factory.create_agent_from_yaml(TOOL_AGENT_YAML)
    response = await agent.run("What's the weather in London?")
    print(f"Agent: {response.text}\n")


# ── D. Workflow from YAML ───────────────────────────────────────────────── #


WORKFLOW_YAML = """\
kind: Workflow
trigger:
  kind: OnConversationStart
  id: research_pipeline
  actions:
    - kind: InvokeAzureAgent
      id: researcher
      agent:
        name: Researcher
      output:
        messages: Local.ResearchOutput
    - kind: InvokeAzureAgent
      id: summarizer
      agent:
        name: Summarizer
      output:
        messages: Local.SummaryOutput
    - kind: EndWorkflow
      id: end
"""


async def workflow_from_yaml() -> None:
    """Create a multi-step workflow from a YAML definition."""
    print("=== D. Workflow from YAML ===\n")

    client = OllamaChatClient(
        host="http://localhost:11434",
        model="qwen3:8b",
    )

    researcher = client.as_agent(
        name="Researcher",
        instructions="You find key facts about a topic. Be thorough but concise.",
    )
    summarizer = client.as_agent(
        name="Summarizer",
        instructions="You summarize research findings into a brief paragraph.",
    )

    factory = WorkflowFactory(
        agents={"Researcher": researcher, "Summarizer": summarizer},
    )

    workflow = factory.create_workflow_from_yaml(WORKFLOW_YAML)
    print("  Workflow created successfully.")
    print(f"  Trigger: {workflow.trigger.kind if hasattr(workflow, 'trigger') else 'N/A'}\n")


# ── E. Round-trip serialization ─────────────────────────────────────────── #


async def round_trip() -> None:
    """Demonstrate agent → YAML → agent round-trip."""
    print("=== E. Round-Trip Serialization ===\n")

    factory = AgentFactory(
        client=OllamaChatClient(
            host="http://localhost:11434",
            model="qwen3:8b",
        ),
        safe_mode=False,
    )

    # Create from YAML.
    agent = factory.create_agent_from_yaml(WEATHER_AGENT_YAML)
    print(f"  Original agent name: {agent.name}")

    # Serialize back (using the definition dict).
    # NOTE: Full YAML round-trip depends on the agent exposing its definition.
    agent2 = factory.create_agent_from_yaml(WEATHER_AGENT_YAML)
    print(f"  Recreated agent name: {agent2.name}\n")


# ── Main ────────────────────────────────────────────────────────────────── #


async def main() -> None:
    await agent_from_yaml_string()
    await agent_from_file()
    await agent_with_tools()
    await workflow_from_yaml()
    await round_trip()


if __name__ == "__main__":
    asyncio.run(main())

File: vignettes/19_declarative/csharp/19_declarative.cs
// Declarative Agents — C#
//
// This sample demonstrates how to define agents and workflows declaratively
// using YAML configuration, then load and run them with the agent factory.
// It mirrors the Julia vignette 19_declarative.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using System.ComponentModel;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// ══════════════════════════════════════════════════════════════════════════ //
//  A. Agent from YAML string                                                //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("=== A. Agent from YAML String ===\n");

const string weatherAgentYaml = """
    kind: Prompt
    name: WeatherHelper
    description: A weather assistant that answers concisely
    instructions: You answer weather questions concisely.
    options:
      temperature: 0.7
      max_tokens: 500
    """;

var chatClient = new OllamaApiClient(new Uri(endpoint), modelName);

// Create an agent factory backed by the Ollama chat client.
var agentFactory = new ChatClientPromptAgentFactory(chatClient);
AIAgent weatherAgent = await agentFactory.CreateFromYamlAsync(weatherAgentYaml);

Console.WriteLine(await weatherAgent.RunAsync("Is it sunny in Paris today?"));
Console.WriteLine();

// ══════════════════════════════════════════════════════════════════════════ //
//  B. Agent from file                                                       //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("=== B. Agent from File ===\n");

var yamlPath = Path.Combine("agents", "weather.yaml");
if (File.Exists(yamlPath))
{
    var yamlContent = await File.ReadAllTextAsync(yamlPath);
    AIAgent fileAgent = await agentFactory.CreateFromYamlAsync(yamlContent);
    Console.WriteLine(await fileAgent.RunAsync("Will it rain in Tokyo tomorrow?"));
}
else
{
    Console.WriteLine($"  Skipped — {yamlPath} not found.");
}
Console.WriteLine();

// ══════════════════════════════════════════════════════════════════════════ //
//  C. Agent with tools                                                      //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("=== C. Agent with Registered Tools ===\n");

[Description("Get the current weather for a location.")]
static string GetWeather(
    [Description("The location to get the weather for.")] string location)
{
    string[] conditions = ["sunny", "cloudy", "rainy", "stormy"];
    var rng = new Random();
    return $"The weather in {location} is {conditions[rng.Next(conditions.Length)]} at {rng.Next(10, 31)}°C.";
}

// Create a separate factory with the tool registered.
var toolAgentFactory = new ChatClientPromptAgentFactory(chatClient, [AIFunctionFactory.Create(GetWeather)]);

const string toolAgentYaml = """
    kind: Prompt
    name: WeatherAgent
    description: Uses tools to answer weather questions
    instructions: Use the available tools to answer weather questions accurately.
    """;

AIAgent toolAgent = await toolAgentFactory.CreateFromYamlAsync(toolAgentYaml);

Console.WriteLine(await toolAgent.RunAsync("What's the weather in London?"));
Console.WriteLine();

// ══════════════════════════════════════════════════════════════════════════ //
//  D. Workflow from YAML (agents wired as a pipeline)                       //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("=== D. Workflow via Declarative Agents ===\n");

const string researcherYaml = """
    kind: Prompt
    name: Researcher
    instructions: You find key facts about a topic. Be thorough but concise.
    """;

const string summarizerYaml = """
    kind: Prompt
    name: Summarizer
    instructions: You summarize research findings into a brief paragraph.
    """;

AIAgent researcher = await agentFactory.CreateFromYamlAsync(researcherYaml);
AIAgent summarizer = await agentFactory.CreateFromYamlAsync(summarizerYaml);

// Chain the two agents: researcher produces text, summarizer condenses it.
var researchResponse = await researcher.RunAsync("Recent advances in quantum computing");
string researchResult = researchResponse.Text;
Console.WriteLine($"[Researcher]: {researchResult}\n");

var summaryResponse = await summarizer.RunAsync($"Summarize the following:\n{researchResult}");
string summary = summaryResponse.Text;
Console.WriteLine($"[Summarizer]: {summary}\n");

// ══════════════════════════════════════════════════════════════════════════ //
//  E. Round-trip serialization                                              //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("=== E. Round-Trip Serialization ===\n");

// Create from YAML, inspect, and recreate.
AIAgent original = await agentFactory.CreateFromYamlAsync(weatherAgentYaml);
Console.WriteLine($"  Original agent: {original.Name}");

// Recreate from the same YAML — demonstrates reproducibility.
AIAgent recreated = await agentFactory.CreateFromYamlAsync(weatherAgentYaml);
Console.WriteLine($"  Recreated agent: {recreated.Name}");

// ══════════════════════════════════════════════════════════════════════════ //
//  F. Environment variable substitution                                     //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("\n=== F. Environment Variable Substitution ===\n");

// In production YAML you would write:
//
//   model:
//     api_key: ${OPENAI_API_KEY}
//     model: gpt-4
//
// The factory resolves ${VAR} references from IConfiguration or
// Environment.GetEnvironmentVariable before parsing.

var apiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
Console.WriteLine(apiKey is not null
    ? "  OPENAI_API_KEY is set — would be substituted in YAML."
    : "  OPENAI_API_KEY is not set — substitution would raise an error.");
